File: qt_cpp_version/QtGeoIp/CustomMarbleWidget.cpp
/* 
 * File:   CustomMarbleWidget.cpp
 * Author: andre
 * 
 * Created on September 24, 2012, 1:25 AM
 */

#include "CustomMarbleWidget.h"
#include <marble/GeoPainter.h>

CustomMarbleWidget::CustomMarbleWidget()
{
    this->marker = NULL; // TODO: DRY
}

CustomMarbleWidget::~CustomMarbleWidget()
{
    if(marker != NULL)
    {
        delete marker;
        marker = NULL;
   }
}

void CustomMarbleWidget::addMarker(const Marker& marker)
{
    this->marker = new Marker;
    this->marker->latitude = marker.latitude;
    this->marker->longitude = marker.longitude;
    this->marker->text = marker.text;
    repaint();
}

void CustomMarbleWidget::customPaint(GeoPainter* painter)
{
        if (marker != NULL)
        {
            painter->setPen(Qt::black);
            GeoDataCoordinates home(
                marker->longitude,
                marker->latitude,
                0.0, GeoDataCoordinates::Degree);
            painter->drawEllipse(home, 7, 7);
            painter->drawText(home, QString::fromStdString(marker->text));
        }
}

File: qt_python_version/custom_marble_widget.py
from PyKDE4.marble import Marble
from PyQt4 import QtCore


class CustomMarbleWidget(Marble.MarbleWidget):
    """
    Extends Marble.MarbleWidget marking it possible to easily add
    markers on the map.
    I've made it that way because MarbleWidget Model seems broken
    in Marble 4.8.5 Python binding.
    """
    def __init__(self):
        super(CustomMarbleWidget, self).__init__()
        self.marker = None

    def addMarker(self, geoDataCoordinates, text):
        self.marker = {
            'geoDataCoordinates': geoDataCoordinates,
            'text': text
        }
        self.repaint()

    def customPaint(self, painter):
        marker = self.marker
        if marker:
            painter.setPen(QtCore.Qt.black)
            home = Marble.GeoDataCoordinates(
                marker['geoDataCoordinates'].longitude(),
                marker['geoDataCoordinates'].latitude(),
                0.0, Marble.GeoDataCoordinates.Degree)
            painter.drawEllipse(home, 7, 7)
            painter.drawText(home, marker['text'])
